File: Leetcode/0076-MinimumWindowSubstring/go/solution.go
func minWindow(s string, t string) string {
    countS := make(map[byte]int)
    countT := make(map[byte]int)
    for i := range t {
        countT[t[i]]++
    }
    // Sliding window
    left, right := 0, 0
    matched := 0
    // Record start and end positions of the substring: [start, end)
    start, end := -1, -1
    for right < len(s) {
        // Expand window
        c := s[right]
        right++
        if _, ok := countT[c]; ok {
            countS[c]++
            if countS[c] == countT[c] {
                matched++
            }
        }
        // Shrink window
        for matched == len(countT) {
            // Update result
            if start == -1 || right-left < end-start {
                start = left
                end = right
            }

            d := s[left]
            left++
            if _, ok := countT[d]; ok {
                if countS[d] == countT[d] {
                    matched--
                }
                countS[d]--
            }
        }
    }
    if start == -1 {
        return ""
    }
    return s[start:end]
}

File: Leetcode/0076-MinimumWindowSubstring/java/Solution.java
class Solution {
    public String minWindow(String s, String t) {
        // Base case
        if (s.length() < t.length()) {
            return "";
        }

        int[] countT = new int[128];
        int matched = 128;
        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            countT[c]++;
            if (countT[c] == 1) {
                matched--;
            }
        }
        // Sliding window
        int start = -1, end = s.length();
        int[] countS = new int[128];
        int left = 0, right = 0;
        while (right < s.length()) {
            char c = s.charAt(right);
            right++;
            if (countT[c] > 0) {
                countS[c]++;
                if (countS[c] == countT[c]) {
                    matched++;
                }
            }
            while (matched == 128) {
                if (right-left < end-start) {
                    end = right;
                    start = left;
                }
                char d = s.charAt(left);
                left++;
                if (countT[d] > 0) {
                    if (countS[d] == countT[d]) {
                        matched--;
                    }
                    countS[d]--;
                }
            }
        }
        if (start == -1) {
            return "";
        }
        return s.substring(start, end);
    }
}